cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/ArtifactIterator.java

```diff
@@ -0,0 +1,123 @@
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.inventory.storage.fs;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.storage.StorageMetadata;
+
+/**
+ * Class that will set the artifactURIs for items returned by a file
+ * system iterator.
+ * 
+ * @author majorb
+ */
+public class ArtifactIterator extends FileSystemIterator {
+    
+    private static final Logger log = Logger.getLogger(ArtifactIterator.class);
+
+    /**
+     * ArtifactIterator constructor.
+     * 
+     * @param dir The directory to iterate
+     * @param ignoreDepth The depth of directories to navigate until non-bucket
+     *     directories are seen.
+     * @param fixedParentDir A path to add to the start of all returned files.
+     * @throws IOException If there is a problem with file-system interaction.
+     */
+    public ArtifactIterator(Path dir, int ignoreDepth, String fixedParentDir) throws IOException {
+        super(dir, ignoreDepth, fixedParentDir);
+        // TODO Auto-generated constructor stub
+    }
+    
+    /**
+     * Get the next file element and add the artifact URI.
+     * @return The next file in the iterator, identified by StorageMetadata.
+     */
+    @Override
+    public StorageMetadata next() {
+        StorageMetadata meta = super.next();
+        URI storageID = meta.getStorageLocation().getStorageID();
+        String ssp = storageID.getSchemeSpecificPart();
+        int firstSlash = ssp.indexOf("/");
+        if (firstSlash < 1) {
+            log.debug("unrecognized storageID format");
+        } else {
+            StringBuilder sb = new StringBuilder();
+            sb.append(ssp.substring(0, firstSlash));
+            sb.append(ssp.substring(firstSlash + 1));
+            meta.artifactURI = URI.create(sb.toString());
```

Comment:
Presumably if URI.create throws an exception it is a bug in this library... should probably use "new URI" and then

```
catch (URISyntaxException ex) {
    throw new RuntimeException("BUG: failed to reconstruct Artifact.uri for " + meta, ex);
}
```